pil-analyzer/src/type_inference.rs

```diff
@@ -0,0 +1,675 @@
+        e: &mut Expression<T>,
+        type_var_mapping: &HashMap<String, Type>,
+    ) -> Result<(), String> {
+        match e {
+            Expression::Number(n, annotated_type) => match annotated_type {
+                Some(TypeName::Int) | Some(TypeName::Fe) | Some(TypeName::Expr) => {}
+                Some(TypeName::TypeVar(tv)) => {
+                    let mut ty = Type::TypeVar(tv.clone());
+                    // Apply regular substitution obtained from unification.
+                    self.substitute(&mut ty);
+                    if !ty
+                        .contained_type_vars()
+                        .all(|tv| type_var_mapping.contains_key(tv))
+                    {
+                        Err(format!("Unable to derive concrete type for literal {n}."))?;
+                    }
+                    // Rename type vars (hopefully just a single one) to match the declaration scheme.
+                    ty.substitute_type_vars(type_var_mapping);
+                    if let Type::TypeVar(tv) = ty {
+                        *annotated_type = Some(TypeName::TypeVar(tv.clone()));
+                    } else {
+                        match ty {
+                            Type::Int => *annotated_type = Some(TypeName::Int),
+                            Type::Fe => *annotated_type = Some(TypeName::Fe),
+                            Type::Expr => *annotated_type = Some(TypeName::Expr),
+                            t => panic!("Invalid resolved type literal number: {t}"),
+                        }
+                    }
+                }
+                _ => panic!("Invalid annotation for literal number."),
+            },
+            Expression::Reference(Reference::Poly(PolynomialReference {
+                name,
+                poly_id: _,
+                generic_args,
+            })) => {
+                for ty in generic_args {
+                    // Apply regular substitution obtained from unification.
+                    self.substitute(ty);
+                    // Now rename remaining type vars to match the declaration scheme.
+                    // The remaining type vars need to be in the declaration scheme.
+                    if !ty
+                        .contained_type_vars()
+                        .all(|tv| type_var_mapping.contains_key(tv))
+                    {
+                        Err(format!(
+                            "Unable to derive concrete type for reference to generic symbol {name}"
+                        ))?;
+                    }
+                    ty.substitute_type_vars(type_var_mapping);
+                }
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    /// Type-checks the isolated expressions.
+    fn check_expressions<T: FieldElement>(
+        &mut self,
+        expressions: &mut [(&mut Expression<T>, ExpectedType)],
+    ) -> Result<(), String> {
+        for (e, expected_type) in expressions {
+            if expected_type.allow_array {
```

Comment:
can you clarify a bit why `.allow_array` is needed and where it comes from?

Reply:
Should be documented at the type: It says if arrays of the expected type are allowed as well. This is mainly used at statement context where we expect either `constr` or `constr[]`.